### ISSUE TITLE
constructors: unable to disambiguate between overloaded methods: one method has no parameters, second has two optional

### ISSUE THREAD

**Issue by FullValueRider** [WaynePhillipsEA/twinbasic#217]
The error message is

unable to disambiguate between overloaded methods in call to 'New'.  2 incompatible matches found:
```
   _AutoKeyByString.New ( Optional ByVal ipFirstKey As String = AutoKeyAid.DefaultKeyForStringKey, Optional ByVal ipKeyCharSequence As String = AutoKeyAid.DefaultKeyCharSequence )
   _AutoKeyByString.New (  )
```

The calling code is
```
Dim myAutoKey As AutoKeyByString  = new AutoKeyByString(p.FirstKey, Join(p.KeyCharSequence.ToArray, vbNullString))
```

**Comment by WaynePhillipsEA** [comment]
This is happening because the p.FirstKey is a Variant, and with passing that Variant argument to the String parameter, twinBASIC considers it a narrowing operation.  For the purposes of overload resolution, narrowing operations are considered to be incompatible.  If you wrap that p.FirstKey in a CStr(), it should match.

**Comment by bclothier** [comment]
I want to comment that this seems to me the problem is really with the fact that you have 2 constructors that can be called with same syntax....

```
Dim myAutoKey As AutoKeyByString  = new AutoKeyByString()
```

How is the compiler supposed to know that you are calling the `New()` and not the `New ( Optional ByVal ipFirstKey As String = AutoKeyAid.DefaultKeyForStringKey, Optional ByVal ipKeyCharSequence As String = AutoKeyAid.DefaultKeyCharSequence )`?

This would not be legal in C# and I wouldn't expect this to be in tB, either. In fact, this can be simplified to just the overloaded parameter with defaults.

**Comment by WaynePhillipsEA** [comment]
@bclothier Sure, it's ambiguous when no arguments are given.   I suspect @FullValueRider only added the no-parameter version just to satisfy the tB requirement of COM exposed classes requiring a default (parameterless) constructor, and tB won't consider the one with optionals as satisfying that requirement at the moment, but it should.

**Comment by FullValueRider** [comment]
@WaynePhillipsEA spot on.  Of course, having thought about this further the overloaded version resolves to three constructors once the constructor parameter types are correctly defined and the conversion to an array list brought to the parameter list from within the body of the method.

- New()
- New(ByVal ipFirstKey as string)
- New(Byval ipKeyCharSequence as ArrayList, Optional byVal ipFirstKey as string )

In the original post the optionals were provided with default parameters which to my mind allows the two constructors to be differentiated because there would never be a case where the optional values would not be provided, there would always be two parameters.

**Comment by FullValueRider** [comment]
Changing to the definitions above eliminates the error flagged at the first line of the module but leaves an error when the class is newe'ed

See AutoKeyByString Class - Clone Method

Project attached

[twinBASIC_VBALib.zip](https://github.com/WaynePhillipsEA/twinbasic/files/6649611/twinBASIC_VBALib.zip)

**Comment by WaynePhillipsEA** [comment]
@FullValueRider that's a narrowing problem again.  All the method calls on ArrayList are late bound, meaning that the return value from your Clone call is a Variant.  Technically the Variant can convert to either of your ArrayList or String overloads, but both are narrowing conversions (from a Variant), and so tB marks them as incompatible (for overload purposes).

**Comment by FullValueRider** [comment]
Ok.  So basically variants as parameters are likely to lead to confusion.  Ditto return values.  That helps in clarifying what I should be doing to use overloading.  Cheers.

**Comment by WaynePhillipsEA** [comment]
No problem.  I'll close this for now, feel free to reopen if necessary.  At some point I'll write an article on narrowing vs widening to explain the overload resolution process more clearly.  It works similarly to how Option Strict mode will work in the future.

**Comment by bclothier** [comment]
Just wondering -- given that a `Variant` parameter type is quite loose-goosey, does it make sense to make it a rule to disallow overloading where the difference is a `Variant` vs. other types? 

Example:

```
Public Sub Foo(Bar As String)
End Sub

Public Sub Foo(Bar As Variant)
End Sub
```

This currently builds but IMO, should be a compile-time error.

**Comment by FullValueRider** [comment]
Variants are deeply embedded in the philosophy of VBA.  For this reason I'd argue the opposite case.  Remove Generics from twinbasic (or at least provide them only to simplify importing VB.NET code) and focus instead on how to provide appropriate compile time/runtime checks on what is and is not allowed when a variant is used.

**Comment by bclothier** [comment]
I think that would be a great loss.

Traditionally, `Variant` has been used as a poor man's overloading, something like this:

```
Public Sub Foo(CouldBeStringOrNumber As Variant)
```

which in turns require runtime checks to verify that the parameter is in fact a string or a number and react accordingly.

In constrast:
```
Public Sub Foo(Bar As String)
...
Public Sub Foo(Bar As Long)
```

requires no runtime checks and is much easier to work. For that reason, I'd prefer compile-time-validated generics over variants with potential error. Because those pose a problem when marshaling across the COM boundary, they wouldn't be exposed anyway. I see two options:

1) we manually create a `Public Sub Foo(CouldBeStringOrNumber As Variant)` which in turn calls the either the `FooImpl(Bar As String)` or `FooImpl(Bar As Long)` or throw an error.

2) the tB compiler can convert the overloaded functions into a single function using `Variant`s, add the necessary runtime checks and throwing an appropriate HRESULT when the type isn't the subtype expected. That way, the tB code has less boilerplate code.

All things being equal, compile time validations ought to be preferred.

**Comment by FullValueRider** [comment]
I totally agree on compile time checking.  My suggestion has been  to allow constraints  for the Types that can be passed by a variant.  Generics, in whatever language, are a sticky plaster to cover over the fact that the type system is inadequate for the task at hand.  I'll happily use generics if that is all that is available.  In VBA, the availability of Variants means that alternative solutions to Generics can be imagined and investigated.

**Comment by mansellan** [comment]
How did this become a question about generics? The suggestion was to restrict overloading, not generics.

Personally I can't see much of an issue with the current arrangement - it's no different to provide an `int` and `object` overload in C#. So long as the compiler is smart enough to select the closest-fitting overload, it shouldn't be a problem?
